File: 2026/06-June/05-Friday/solution.kt
class Solution {

    data class PairResult(
        val count: Long,
        val waviness: Long
    )

    private lateinit var s: String
    private val memo = HashMap<String, PairResult>()

    private fun dfs(
        pos: Int,
        prev: Int,
        prev2: Int,
        lead: Boolean,
        tight: Boolean
    ): PairResult {

        if (pos == s.length) {
            return PairResult(1L, 0L)
        }

        val key = "$pos|$prev|$prev2|$lead"

        if (!tight && memo.containsKey(key)) {
            return memo[key]!!
        }

        val limit =
            if (tight) s[pos] - '0'
            else 9

        var cnt = 0L
        var wav = 0L

        for (d in 0..limit) {

            val ntight = tight && d == limit
            val nlead = lead && d == 0

            val res = dfs(
                pos + 1,
                if (nlead) -1 else d,
                prev,
                nlead,
                ntight
            )

            cnt += res.count

            if (!lead &&
                prev2 != -1 &&
                (
                    (prev2 < prev && prev > d) ||
                    (prev2 > prev && prev < d)
                )
            ) {
                wav += res.count
            }

            wav += res.waviness
        }

        val ans = PairResult(cnt, wav)

        if (!tight) {
            memo[key] = ans
        }

        return ans
    }

    private fun solve(x: Long): Long {

        if (x <= 0L) return 0L

        s = x.toString()
        memo.clear()

        return dfs(
            0,
            -1,
            -1,
            true,
            true
        ).waviness
    }

    fun totalWaviness(
        num1: Long,
        num2: Long
    ): Long {

        return solve(num2) - solve(num1 - 1)
    }
}

File: 2026/06-June/05-Friday/solution.cs
public class Solution
{
    string s;
    Dictionary<string,(long,long)> memo;

    (long,long) DFS(
        int pos,
        int prev,
        int prev2,
        bool lead,
        bool tight)
    {
        if(pos==s.Length)
            return (1,0);

        string key =
            pos + "|" + prev + "|" +
            prev2 + "|" + lead;

        if(!tight && memo.ContainsKey(key))
            return memo[key];

        int limit = tight ? s[pos]-'0' : 9;

        long cnt = 0;
        long wav = 0;

        for(int d=0; d<=limit; d++)
        {
            bool ntight = tight && d==limit;
            bool nlead = lead && d==0;

            var res = DFS(
                pos+1,
                nlead ? -1 : d,
                prev,
                nlead,
                ntight
            );

            cnt += res.Item1;

            if(!lead &&
               prev2!=-1 &&
               ((prev2<prev && prev>d) ||
                (prev2>prev && prev<d)))
                wav += res.Item1;

            wav += res.Item2;
        }

        if(!tight)
            memo[key]=(cnt,wav);

        return (cnt,wav);
    }

    long Solve(long x)
    {
        if(x<=0) return 0;

        s=x.ToString();
        memo=new Dictionary<string,(long,long)>();

        return DFS(0,-1,-1,true,true).Item2;
    }

    public long TotalWaviness(long num1,long num2)
    {
        return Solve(num2)-Solve(num1-1);
    }
}
